File: deleteandEarn.java
/*
Time Complexity -->
O(n)

Space Complexity-->
O(1)

Approach--.
Similar to house robber where we can choose between the adjacents and check for the the maximum. 
*/

class Solution {
    public int deleteAndEarn(int[] nums) {
        int max = nums[0];

        for (int i = 1; i < nums.length; i++) {

      // compare all elements with max
      // assign maximum value to max
        max = Math.max(max, nums[i]);

        }
        System.out.println(max);
        int arr[] = new int[max+1];
        
        //  for(int j=0;j<arr.length;j++){
        //     System.out.println(arr[]);
        // }//for
        
        for(int i=0;i<nums.length;i++){// count number of frequency
            arr[nums[i]]+=1;
        }//for
                       
        // for(int j=0;j<arr.length;j++){
        //     System.out.println(arr[j]);
        // }//for
        
        int nottake = 0;
        int take = arr[0];
        int tmp = 0;
        for(int i=1;i<arr.length;i++){
            tmp = nottake;
            nottake = Math.max(nottake, take);
            take = tmp+arr[i]*i;
        }
        
        return Math.max(nottake, take);
    }
}
                       

File: Delete-And-Earn.py
# Time Complexity: O(n) + O(max(n)) were n is the size of nums and max(n) is maximum value of the
# Space Complexity: O(max(n))

class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        size = max(nums)
        dp = [0 for i in range(size+1)]

        for i in nums:
            dp[i] += i

        skip = 0
        take = dp[0]

        for i in range(1, len(dp)):
            temp = skip
            skip = max(skip, take)
            take = temp + dp[i]
        return max(skip, take)